flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/ParserImpl.java

```diff
@@ -95,12 +105,24 @@ public ParserImpl(
         }
 
         // parse the sql query
-        SqlNode parsed = parser.parse(statement);
-
-        Operation operation =
-                SqlToOperationConverter.convert(planner, catalogManager, parsed)
-                        .orElseThrow(() -> new TableException("Unsupported query: " + statement));
-        return Collections.singletonList(operation);
+        try {
+            // use parseSqlList here because we need to support statement end with ';' in sql
+            // client.
+            SqlNodeList sqlNodeList = parser.parseSqlList(statement);
+            List<SqlNode> parsed = sqlNodeList.getList();
+            Preconditions.checkArgument(parsed.size() == 1, "only single statement supported");
+            return Collections.singletonList(
+                    SqlToOperationConverter.convert(planner, catalogManager, parsed.get(0))
+                            .orElseThrow(
+                                    () -> new TableException("Unsupported query: " + statement)));
+        } catch (SqlParserException parserException) {
+            if (config.getConfiguration().getBoolean(TABLE_PARSER_ALLOW_PARTIAL_PARSE)
+                    && parserException.getMessage().contains("Encountered \"<EOF>\"")) {
+                return Collections.singletonList(new StagingOperation(statement, parserException));
+            } else {
+                throw parserException;
+            }
```

Comment:
I would rather implement this differently: More than having an operation here that wraps the parsing exception etc, which then is used to properly handle EOF by the client, I would instead have a new method in the parser that helps with this "interactive parsing" mode.

For example, it could be some method like `interactiveParse` which returns a union type: either the operation or the parse failure. No need to use the operation stack for that, as it's really counterintuitive. This new method will also allow you to get rid of this internal config option and simplify the implementation overrall.

Also, the check based on string seems a bit fragile. Can we have a more robust check of some kind? Is there some sub exception that calcite uses?

Reply:
thanks for the advise, I am also thinking about how to get rid of the internal config based on the Godfreyhe 's comment. I like the idea adding a new function, but I would like to keep the StagingOperation instead of a union type, because it make it easier to transporting the result between different components in parser/client. 

I am afraid that we can only check based on string, because the calcite parser only format the message to describe the error, there is no sub exception.

WDYT?